File: app/external.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any


class DependencyError(RuntimeError):
    pass


class TransientDependencyError(DependencyError):
    pass


@dataclass(frozen=True)
class DependencyOutcome:
    outcome: str
    message: str
    reference: str


class ExternalDependencyGateway:
    def evaluate(self, dependency_name: str, payload: dict[str, Any], attempt_number: int) -> DependencyOutcome:
        if dependency_name != "fraud_service":
            raise DependencyError(f"Unsupported dependency '{dependency_name}'.")

        entity_id = str(
            payload.get("applicant_id")
            or payload.get("claim_id")
            or payload.get("vendor_id")
            or payload.get("employee_id")
            or payload.get("document_id")
            or "unknown"
        )
        dependency_mode = payload.get("dependency_mode", "pass")

        if dependency_mode == "transient_error" and attempt_number == 0:
            raise TransientDependencyError("Simulated transient dependency timeout from fraud_service.")

        if dependency_mode == "fail" or payload.get("fraud_flag") is True:
            return DependencyOutcome(
                outcome="fail",
                message="fraud_service flagged the request as suspicious.",
                reference=f"fraud_service:{entity_id}:fail",
            )

        return DependencyOutcome(
            outcome="pass",
            message="fraud_service cleared the request.",
            reference=f"fraud_service:{entity_id}:pass",
        )

File: web/lib/external.ts
export class DependencyError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "DependencyError";
  }
}

export class TransientDependencyError extends DependencyError {
  constructor(message: string) {
    super(message);
    this.name = "TransientDependencyError";
  }
}

export interface DependencyOutcome {
  outcome: string;
  message: string;
  reference: string;
}

export class ExternalDependencyGateway {
  evaluate(dependencyName: string, payload: Record<string, unknown>, attemptNumber: number): DependencyOutcome {
    if (dependencyName !== "fraud_service") {
      throw new DependencyError(`Unsupported dependency '${dependencyName}'.`);
    }

    const entityId = String(
      payload["applicant_id"] ??
        payload["claim_id"] ??
        payload["vendor_id"] ??
        payload["employee_id"] ??
        payload["document_id"] ??
        "unknown"
    );

    const dependencyMode = String(payload["dependency_mode"] ?? "pass");

    if (dependencyMode === "transient_error" && attemptNumber === 0) {
      throw new TransientDependencyError("Simulated transient dependency timeout from fraud_service.");
    }

    if (dependencyMode === "fail" || payload["fraud_flag"] === true) {
      return {
        outcome: "fail",
        message: "fraud_service flagged the request as suspicious.",
        reference: `fraud_service:${entityId}:fail`,
      };
    }

    return {
      outcome: "pass",
      message: "fraud_service cleared the request.",
      reference: `fraud_service:${entityId}:pass`,
    };
  }
}
